packages/amplify-graphql-sql-transformer/src/graphql-sql-transformer.ts

```diff
@@ -0,0 +1,148 @@
+  }
+
+  field = (
+    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
+    definition: FieldDefinitionNode,
+    directive: DirectiveNode,
+    ctx: TransformerSchemaVisitStepContextProvider,
+  ): void => {
+    if (parent.name.value !== 'Query' && parent.name.value !== 'Mutation') {
+      throw new InvalidDirectiveError(
+        `@sql directive can only be used on Query or Mutation types. Check type "${parent.name.value}" and field "${definition.name.value}".`,
+      );
+    }
+
+    const directiveWrapped = new DirectiveWrapper(directive);
+    if (
+      !directive?.arguments?.find((arg) => arg.name.value === 'statement') &&
+      !directive?.arguments?.find((arg) => arg.name.value === 'reference')
+    ) {
+      throw new InvalidDirectiveError(
+        `@sql directive must have either a 'statement' or 'reference' argument. Check type "${parent.name.value}" and field "${definition.name.value}".`,
+      );
+    }
+
+    const args = directiveWrapped.getArguments(
+      {
+        resolverTypeName: parent.name.value,
+        resolverFieldName: definition.name.value,
+      } as SqlDirectiveConfiguration,
+      generateGetArgumentsInput(ctx.transformParameters),
+    );
+    let resolver = this.sqlDirectiveFields.get(definition);
+
+    if (resolver === undefined) {
+      resolver = [];
+      this.sqlDirectiveFields.set(definition, resolver);
+    }
+
+    resolver.push(args);
+  };
+
+  generateResolvers = (context: TransformerContextProvider): void => {
+    if (this.sqlDirectiveFields.size === 0) {
+      return;
+    }
+
+    const stack: cdk.Stack = context.stackManager.createStack(SQL_DIRECTIVE_STACK);
+    const env = context.synthParameters.amplifyEnvironmentName;
+
+    stack.templateOptions.templateFormatVersion = '2010-09-09';
+    stack.templateOptions.description = 'An auto-generated nested stack for the @sql directive.';
+
+    new cdk.CfnCondition(stack, ResourceConstants.CONDITIONS.HasEnvironmentParameter, {
+      expression: cdk.Fn.conditionNot(cdk.Fn.conditionEquals(env, ResourceConstants.NONE)),
+    });
+
+    this.sqlDirectiveFields.forEach((resolverFns) => {
+      resolverFns.forEach((config) => {
+        const { RDSLambdaDataSourceLogicalID: dataSourceId } = ResourceConstants.RESOURCES;
+        const dataSource = context.api.host.getDataSource(dataSourceId);
+        const MISSING_QUERY_STATEMENT = 'MISSING_CUSTOM_QUERY';
+        if (config.reference && !context.customQueries.has(config.reference)) {
+          throw new InvalidDirectiveError(
+            `@sql directive 'reference' argument must be a valid custom query name. Check type "${config.resolverTypeName}" and field "${config.resolverFieldName}". The custom query "${config.reference}" does not exist in "sql-statements" directory.`,
+          );
+        }
+
+        const statement = config.statement ?? (config.reference ? context.customQueries.get(config.reference) : MISSING_QUERY_STATEMENT);
```

Comment:
Instead of falling back to a `MISSING_QUERY_STATEMENT` and proceeding with resolver generation, shouldn't the case of not having valid `statement` or a pre-loaded `reference` SQL file be an Error instead? If you already have those checks, is `MISSING_QUERY_STATEMENT` default needed at all?

Reply:
MISSING_QUERY_STATEMENT has been added to make the compiler happy. Otherwise `statement` would be `string | undefined` instead of a `string`. I could have used `context.customQueries.get(config.reference)!` but preferred this approach better.

Reply:
There are couple of cases that we should check: 
- When the `statements` here is an empty string.
- When both statements and reference are provided.